compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/SingletonReferencesLowering.kt

```diff
@@ -52,7 +52,10 @@ private class SingletonReferencesLowering(val context: JvmBackendContext) : File
     }
 
     override fun visitGetObjectValue(expression: IrGetObjectValue): IrExpression {
-        val instanceField = context.declarationFactory.getFieldForObjectInstance(expression.symbol.owner)
+        val instanceField = if (allScopes.any { it.irElement == expression.symbol.owner })
```

Comment:
Looks like `$$INSTANCE` field should be used only in companion <clinit>, otherwise it could be exposed by inline through synthetic accessor:
```
interface Test {
    companion object {
        fun ok() = "OK"
        
        inline fun test() = run {
            Test.ok()
        }
    }
}
```

Reply:
Nice catch, in this case it should in fact be the dispatch receiver of `test`, so neither the original nor the changed code is entirely correct.

Reply:
...though it actually works with `$$INSTANCE` as well because an accessor for it is created.